cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,205 @@
+	timestampFormat string
+
+	// timeProvider is a function that returns the current time. Typically time.Now. Exposed for testing.
+	timeProvider func() time.Time
+
+	// mutationLock should be taken when reading or writing trace, userAndDebugBuffer, or outputInfo.
+	mutationLock sync.Mutex
 }
 
-func (e *LogEntry) String() string {
-	return fmt.Sprintf("%s %s", e.LocalTimestamp.Format("2006-01-02T15:04:05Z"), e.Message)
+// WriteUser writes message to the underlying log.Logger, and then buffers the message
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteUser(message string) {
+	l.write(l.userPrefix, message)
 }
 
-// LoggerInterface is logger abstraction
-type LoggerInterface interface {
-	Log(message string)
+// WriteDebug writes message to the underlying log.Logger, and then buffers the message
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteDebug(message string) {
+	l.write(l.debugPrefix, message)
 }
 
-// NewStdoutLogger creates a new logger which uses prefix for all messages logged.
-// All messages are sent to stdout.
-func NewStdoutLogger(prefix string) LoggerInterface {
-	return stdoutLogger{prefix: prefix}
+// WriteTrace buffers the message for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteTrace(message string) {
+	l.mutationLock.Lock()
+	defer l.mutationLock.Unlock()
+
+	l.trace = append(l.trace, message)
 }
 
-type stdoutLogger struct{ prefix string }
+// WriteMetric keeps non-nil fields from metric for inclusion in ReadOutputInfo().
+// Elements of list fields are appended to the underlying view.
+func (l *defaultToolLogger) WriteMetric(metric *pb.OutputInfo) {
+	l.mutationLock.Lock()
+	defer l.mutationLock.Unlock()
 
-// Log logs a message
-func (l stdoutLogger) Log(message string) {
-	fmt.Printf("%s %s\n", l.prefix, newLogEntry(message))
+	proto.Merge(l.outputInfo, metric)
 }
 
-// NewDefaultLogger creates a new logger that sends all messages to the default logger.
-func NewDefaultLogger() LoggerInterface {
-	return defaultLogger{}
+// Returns a view comprised of:
+//   - Calls to WriteMetric
+//   - All user, debug, and trace logs. User and debug logs are appended into a single
+//     member of OutputInfo.SerialLogs; each trace log is a separate member.
+// All buffers are cleared when this is called. In other words, a subsequent call to
+// ReadOutputInfo will return an empty object.
+func (l *defaultToolLogger) ReadOutputInfo() *pb.OutputInfo {
+	// Locking since ReadOutputInfo has a side effect of clearing the internal state.
+	l.mutationLock.Lock()
+	defer l.mutationLock.Unlock()
+
+	// Prepend stdout logs (user, debug) to the trace logs.
+	var combinedTrace []string
+	if l.userAndDebugBuffer.Len() > 0 {
+		combinedTrace = []string{l.userAndDebugBuffer.String()}
+		l.userAndDebugBuffer.Reset()
+	}
+	ret := &pb.OutputInfo{SerialOutputs: append(combinedTrace, l.trace...)}
+	l.trace = []string{}
+	proto.Merge(ret, l.outputInfo)
+	l.outputInfo = &pb.OutputInfo{}
```

Comment:
minor: this would be more readable if output building and buffer reset were in two separate functions that are called from here. 
e.g.:
ret := buildOutput()
resetBuffers()
return ret

Reply:
done